### PR TITLE
[FIXED JENKINS-29989] Update RunIdMigrator to properly handle matrix and Maven jobs

### PR DESCRIPTION
**Pull request by ghost**
[JENKINS-29989](https://issues.jenkins-ci.org/browse/JENKINS-29989)

Properly revert job history for matrix and maven jobs.

This should be applied to the 1.609.X LTS branch as well since that is the first LTS to use the new build id format.

### CODE REVIEW

**Comment by jpederzolli** [comment]
Test failures in unrelated code:

hudson.slaves.CommandLauncherTest.commandFails

File: core/src/main/java/jenkins/model/RunIdMigrator.java [hunk 1]
```diff
@@ -350,6 +350,11 @@ private void unmigrateJobsDir(File jobs) throws Exception {
             return;
         }
         for (File job : jobDirs) {
+
+            if (job.getName().equals("builds")) {
```
Comment (jglick) [inline core/src/main/java/jenkins/model/RunIdMigrator.java]: Comment now on line 369 should be edited to clarify that the older clause handles grandchildren from `getRootDirFor` whereas this clause handles direct children.

**Comment by jglick** [comment]
:+1:

The deadline for 1.609.3 has long since passed I am afraid.

**Comment by KostyaSha** [comment]
@jglick why not do 1.609.4?
cc @olivergondza

**Comment by KostyaSha** [comment]
1.609.3 was delayed, why not pick it? Or at least merge into master?

**Comment by KostyaSha** [comment]
Please squash and exclude "minor cleanup to trigger pull request" for re-triggering this damn PR builder you can close and open PR.

**Comment by KostyaSha** [comment]
@jglick you are messing commit history and git-blame with such commits, it not so difficult to squash it.

**Comment by jglick** [comment]
Nor is it so difficult to use the Git commands that pay attention only to the first merge parent. History is history, you can look at it in different ways, but rebasing is a black hole.

**Comment by KostyaSha** [comment]
@jglick project is using github and i'm looking at history using it's web and more and more history looks like a trash, because of somebody laziness. I don't say rebase, i suggested fixup commits into one. Merging such trash https://github.com/jpederzolli/jenkins-1/commit/1c0e94bfe95de6fb9087318bc208c2091f9b2d08 is a shame.

File: core/src/test/java/jenkins/model/RunIdMigratorTest.java [hunk 2]
```diff
@@ -128,6 +128,26 @@
         assertEquals("{1=→2014-01-02_03-04-05, 2014-01-02_03-04-05={build.xml='<?xml version='1.0' encoding='UTF-8'?>\n<run>\n  <stuff>ok</stuff>\n  <number>1</number>\n  <otherstuff>ok</otherstuff>\n</run>'}}", summarize());
     }
 
+    @Test public void reverseMatrixAfterNewBuilds() throws Exception {
+        File root = dir;
+        dir = new File(dir, "jobs/someproject/Environment=prod/builds");
+        write("1/build.xml", "<?xml version='1.0' encoding='UTF-8'?>\n<run>\n  <stuff>ok</stuff>\n  <timestamp>1388649845000</timestamp>\n  <otherstuff>ok</otherstuff>\n</run>");
```
Comment (lanwen) [inline core/src/test/java/jenkins/model/RunIdMigratorTest.java]: please move such strings to resources in future

**Comment by jpederzolli** [comment]
@KostyaSha thanks - I will squash commits as necessary in the future.

**Comment by KostyaSha** [comment]
@jpederzolli thanks for fix!

**Pull request merged**